lib/api/api_test.go

```diff
@@ -1778,3 +1863,1487 @@ func runningInContainer() bool {
+					{
+						ID:            base64.RawURLEncoding.EncodeToString([]byte{1, 2, 3, 4}),
+						RpId:          "localhost",
+						PublicKeyCose: base64.RawURLEncoding.EncodeToString(publicKeyCose),
+						RequireUv:     false,
+					},
+				},
+			},
+		}))
+		baseURL, cancel, _, err := startHTTP(cfg)
+		if err != nil {
+			t.Fatal(err, "Failed to start HTTP server")
+		}
+		t.Cleanup(cancel)
+
+		httpRequest := func(method string, url string, body any) *http.Response {
+			t.Helper()
+			var bodyReader io.Reader = nil
+			if body != nil {
+				bodyBytes, err := json.Marshal(body)
+				if err != nil {
+					t.Fatal(err, "Failed to marshal HTTP request body")
+				}
+				bodyReader = bytes.NewReader(bodyBytes)
+			}
+
+			req, err := http.NewRequest(method, baseURL+url, bodyReader)
+			if err != nil {
+				t.Fatal(err, "Failed to construct HttpRequest")
+			}
+
+			client := http.Client{
+				Timeout: 15 * time.Second,
+			}
+			resp, err := client.Do(req)
+			if err != nil {
+				t.Fatal(err, "Failed to execute HTTP request")
+			}
+
+			return resp
+		}
+
+		httpPost := func(url string, body any) *http.Response {
+			t.Helper()
+			return httpRequest(http.MethodPost, url, body)
+		}
+
+		getAssertionOptions := func() startWebauthnAuthenticationResponse {
+			startResp := httpPost("/rest/noauth/auth/webauthn-start", nil)
+			if startResp.StatusCode != http.StatusOK {
+				t.Fatalf("Failed to start WebAuthn registration: status %d", startResp.StatusCode)
+			}
+			if hasSessionCookie(startResp.Cookies()) {
+				t.Errorf("Expected no session cookie when starting WebAuthn registration")
+			}
+
+			var respBody startWebauthnAuthenticationResponse
+			err := unmarshalTo(startResp.Body, &respBody)
+			if err != nil {
+				t.Fatal(err, "Failed to unmarshal CredentialAssertion")
+			}
+
+			return respBody
+		}
+
+		return httpPost, getAssertionOptions
+	}
+
+	type passwordAuthResponseBody struct {
+		Username     string
+		Password     string
+		StayLoggedIn bool
+	}
+
+	t.Run("Can log in with password instead of WebAuthn", func(t *testing.T) {
+		t.Parallel()
+		httpPost, _ := startServer(t)
+
+		finishResp := httpPost("/rest/noauth/auth/password", passwordAuthResponseBody{
+			Username:     "user",
+			Password:     "räksmörgås",
+			StayLoggedIn: false,
+		})
+		if finishResp.StatusCode != http.StatusNoContent {
+			t.Fatalf("Failed password authentication: status %d", finishResp.StatusCode)
+		}
+	})
+
+	t.Run("Can log in with WebAuthn instead of password", func(t *testing.T) {
+		t.Parallel()
+		httpPost, getAssertionOptions := startServer(t)
+		startResp := getAssertionOptions()
+
+		cred := createWebauthnAssertionResponse(startResp.Options, []byte{1, 2, 3, 4}, privateKey, "https://localhost:8384", false, 1, t)
+
+		finishResp := httpPost("/rest/noauth/auth/webauthn-finish", startResp.finish(&cred, false))
+		if finishResp.StatusCode != http.StatusNoContent {
+			t.Fatalf("Failed WebAuthn authentication: status %d", finishResp.StatusCode)
```

Comment:
We should re-evaluate whether `t.Fatalf(...)` is needed here.  I think most of these could use `t.Errorf(...)`, which would keep testing the other cases.